core/src/main/java/club/mondaylunch/gatos/core/graph/Node.java

```diff
@@ -45,6 +47,18 @@ private Node(
         this.outputs = Map.copyOf(outputs.stream().collect(Collectors.toMap(NodeConnector::name, Function.identity())));
     }
 
+    /**
+     * Needed for deserialization.
+     */
+    @ApiStatus.Internal
+    public Node() {
```

Comment:
Is it making a new node and changing all its final fields with reflection? I really don't like that aesthetically but I guess it works. A custom codec would be nicer, wouldn't violate the expectations of final fields not changing, could let us not have to change the method names, & would let us only serialize the fields needed. I think I understand how codecs work now from reading this PR, I'm happy to make it myself.

Reply:
I've avoided making custom codecs unless absolutely needed to avoid needing to change the codec whenever the class is changed.

Reply:
I can make a custom codec if you would like though.